### DISCUSSION TITLE
Symbols

### DISCUSSION THREAD

**Discussion by lexoyo** [GrapesJS/grapesjs#4317] [Q&A] [answered]
Hello
I am integrating grapesjs in silex :tada: (I will release as much grapesjs plugins as I can)
My question is about symbols, I guess it will be for @artf :pray: 
Do you think I should use [your POC](https://github.com/artf/grapesjs/pull/3165) (and contribute if I can?) or create a separate plugin to manage all syncs?
I guess the only problem with the POC is that the symbol is added to the page as you said in the PR

**Comment by artf** [comment]
Awesome @lexoyo 
I'd say I'm ok with having symbols in the core (maybe it would be worth moving all the related methods outside of the Component to make it more readable) but as you mentioned the main issue right now is not providing the way to store the main symbols properly.

**Reply by lexoyo** [reply]
thank you for your answer :pray: 
it may be easier for me to do a plugin, what do you think?
i'm not familiar with backbone, i can learn but i would have questions
the main symbols should be on a collection in editor.Symbols?

**Reply by artf** [reply]
The backbone is used here mainly as a simple event emitter, so not a big deal. 
I'd say the main symbols should be collected at the Component module level (similar to how all other components are tracked via `editor.Components.componentsById`).
The next step would be to properly store them on the JSON project (via `editor.Components.store/load`).
```js
{
 assets: [...],
 styles: [...],
 pages: [...],
 // New collection
 symbols: [...],
}
```
About the plugin, I still think it might be better to keep this functionality in the core due to a strict relation with the base component and the JSON project data but idk, I guess it might also depend on how well the plugin could play in this scenario.

**Reply by lexoyo** [reply]
Ok thank you for the answer

I started learning backbone, i never thought i would one day :sweat_smile: 

The concept would be this ?

* every time a component is created/deleted, we check if it is a symbol and add/remove it to/from the `Symbol`
* no more "master" component
* every time a component related to a symbol is updated we update the other components of its symbols (2 way sync) or only if it asks us to (1 way sync)?
  
Here is what i understand i should do

1. split the symbols related features from `CssRule.js`, `Component.js` to `Symbol.js`
2. from `Components.js` to `Symbols.js`
3. add the `Symbols` collection (`editor.Symbols`)
4. save the symbols (minus the components which are maintained at runtime)

**Comment by lexoyo** [comment]
I ended up doing it as a plugin and not using your implementation of Symbols @artf :-/

I feel like I need to validate what I am doing as it could be a nice plugin for everyone. Also I definitely can hear a "that is not the way to go", please don't be gentle and let's discuss it with anyone / everyone

**Here is what I did for now**

On editor `load`

* `editor.Symbols = new Symbols([], { options, editor })`, this is a collection of symbols, each symbol has a `components` list
* `new SymbolsView(editor.Symbols, { options, editor })`, this is the list of symbols, draggable to the stage
* `Storage(editor, opts)` => adds the symbols key to the saved/loaded data (saved data does not contain the symbols `components` lists)

On `storage:end:load` I browse all components and update all the symbols `components` lists

On `component:create` and `component:remove` I maintain the symbols `components` lists up to date

On drop from the list I create a new element with the symbols `json`

[The WIP code is here](https://github.com/silexlabs/grapesjs-sympbols)

The list of symbols looks like this, with the number of components for each symbol:

![WIP symbols plugin](https://user-images.githubusercontent.com/715377/168800151-b75f2fa9-1cdc-4655-a981-f98a6e1d3d03.png)

**Here is my todo list / what I plan to do next**

EDIT: [todo list moved here](https://github.com/lexoyo/grapesjs-sympbols/issues/1#issuecomment-1139848294)

**Reply by artf** [reply]
Looks good to me Alex, I'm actually curious about seeing Symbol implementation integrated via a plugin. I'm not sure if the editor is able to provide all the necessary events but in that case, let me know.

**Reply by lexoyo** [reply]
Hello

I'm back at the symbol plugin, with new ideas and questions concerning the events:

1. would you know if this is correct?
    | to be synced | event to listen to |
    | -- | -- |
    | add/remove children to a symbol or to a child | `component:update:components` |
    | classes on a symbol's child | `component:update:classes` |
    | text content of a symbol or child | `component:input` |
    | attributes of a symbol or child | `component:update:attributes` |
1. when I **drag** a **new** component over a component on the stage, I see the event `component:update:components` fired 2 times, is it intended, any pointer on how to detect that a component has been actually dropped ?

**Reply by lexoyo** [reply]
FYI: I updated [the "roadmap"](https://github.com/silexlabs/grapesjs-symbols/issues/1#issuecomment-1139848294) with my progress, take a look at the demo to see how it is working so far
https://codepen.io/lexoyo/pen/xxJGEwo

cc @artf

**Reply by lexoyo** [reply]
Maybe easier with a video
https://youtu.be/hR0GXWF-_vc

**Reply by artf** [reply]
Hi @lexoyo sorry for the late reply, thanks for updating about your progress, it looks really good. 
I can confirm about the list of events and also the intention of  `component:update:components` trigger during new component drag. It's used to create (first event) and remove (second event) a temporary component from the block content in order to understand where the component can be dropped. 
I think in the future the internal creation action in the Sorter will be replaced by [this method](https://github.com/artf/grapesjs/blob/17f0c4f62c4867f164e548200869929654010693/src/dom_components/index.ts#L601), but for now, `component:update:components` provides options containing information about temporary actions.

**Reply by lexoyo** [reply]
Thank you @artf !
A question for when you will have time: when I sync the changes of a text element, the cursor is sent back to the beginning of the element. Do you know why? Is it normal? Is it on grapesjs side or on my side maybe?
I'll investigate, just asking in case you know this behavior

**Reply by lexoyo** [reply]
> Thank you @artf ! A question for when you will have time: when I sync the changes of a text element, the cursor is sent back to the beginning of the element. Do you know why? Is it normal? Is it on grapesjs side or on my side maybe? I'll investigate, just asking in case you know this behavior

this is still a problem :)

**Comment by lexoyo** [comment]
I'm very happy about how it goes for now, grapesjs is really awesome @artf 

The synchro of instances between pages works now and the symbols in symbols too

Would anyone test it and let me know what bugs they have?

Also I put many `TODO: Needs review` in specific places, in case you have time @artf : https://github.com/silexlabs/grapesjs-symbols/search?q=%22TODO%3A+needs+review%22

:rocket:

**Reply by artf** [reply]
Awesome work @lexoyo ❤️ I'll try to find time to review it

**Reply by lexoyo** [reply]
I'm still working on it and it still have several important bugs, so no hurry, but i'd apreciate it greatly at one point

**Reply by lexoyo** [reply]
bump @artf for the review, but I understand if it is too much to ask :pray: 

also I have a question: I don't find an event to detect a component style change (i tried component, component:update, component:update:style, component:styleUpdate, change, style, selector)

**Reply by lexoyo** [reply]
> I have a question: I don't find an event to detect a component style change (i tried component, component:update, component:update:style, component:styleUpdate, change, style, selector)

In the end I sync "inline" styles on `styleable:change`: https://github.com/silexlabs/grapesjs-symbols/blob/main/src/model/Symbols.js#L145

**Comment by lexoyo** [comment]
Another question, this is probably stupid but I struggle with carret position

When I change the text content of a component in a symbol, I update all other occurences with `occurence.components(changed.getCurrentView().getContent())` and this seems to reset the carret position in `changed` to `0`

I can't get the carret position with `window.getSelection()`, because never has any range in the callback of `editor.on('component:input'`

Would anyone have a pointer?

**Comment by lexoyo** [comment, marked as answer]
Hello  @artf
I guess you're busy and that's very ok, I just hope you're fine
I figured everything out, all the important and touchy features are implemented
[DEMO](https://codepen.io/lexoyo/full/xxJGEwo)
[Repo](https://github.com/silexlabs/grapesjs-symbols/)
I will battle test it on Silex soon
Thank you for your help!!

**Reply by artf** [reply]
Sorry Alex, unfortunately it's a bit an hectic period right now so can't really find time to check your implementation but I'm happy to hear you were able to figure it out by yourself.

**Reply by lexoyo** [reply]
No worries!

**Comment by Samkanda** [comment]
Hey, how would you import this package in react(nextjs). I am getting a module not found when I attempt it. Issue is in error tab as well with screen shot

**Reply by lexoyo** [reply]
Hi I am not sure I understand, would you open an issue here and explain what is the error etc ?
https://github.com/silexlabs/grapesjs-symbols/issues/new

**Comment by rokon-rabbi** [comment]
hello! i want to use html and css code that is generated and with that i want to show new webpage in another tab. how to solve this?

**Reply by lexoyo** [reply]
Hello
Not sure what you mean but symbols are not useful accross tabs
Please create another issue
